### ISSUE TITLE
Printing of and error messages involving Records, Structures and Variants + doc of Set Printing All

### ISSUE THREAD

**Issue by Zimmi48** [Zimmi48/bugzilla-test#5303]
Note: the issue was created automatically with bugzilla2github tool

# Bugzilla Bug ID: 5303
Date: 2017-01-10 15:35:13 +0100
From: @_amahboubi
To:   
CC:   coq-bugs-redist@lists.gforge.inria.fr, @_ejgallego, @_herbelin

Last updated: 2017-01-16 15:52:52 +0100

## Bugzilla Comment ID: 15076
Date: 2017-01-10 15:35:13 +0100
From: @_amahboubi

Created attachment 801
Dummy examples of defintion of records/variant types, with printing queries

I must confess I discovered only recently the existence of "Variant". 

The attached session illustrates some situations that I find confusing, both in the wording of error messages and in the display of the types in response to a Print query in Set Printing All mode.

Also, if I am not mistaken, it illustrates that the documentation of the "Set Printing All" command is incomplete.

> Attached file: records_variants_structures.v (text/x-verilog, 2723 bytes)
> Description:   Dummy examples of defintion of records/variant types, with printing queries

## Bugzilla Comment ID: 15080
Date: 2017-01-10 22:52:14 +0100
From: @_ejgallego

Interesting examples, I guess the printing problems should come from externalization, but I am not sure I understand how inductive definitions are "externalized".

The raw datatype is this:

> ((mind_packets
>  (((mind_typename (Id snatbox))
>   (mind_arity_ctxt ())
>   (mind_arity
>     (RegularArity ((mind_user_arity (Sort (Prop Pos))) (mind_sort (Prop Pos)))))
>   (mind_consnames ((Id Snatbox)))
>   (mind_user_lc
>     ((Prod (Name (Id smynat))
>            (Ind (((Mutind (MPfile (DirPath ((Id Datatypes) (Id Init) (Id Coq))))
>                 (DirPath ()) (Id nat)) 0) (Instance ())))
>            (Rel 2))))
>   (mind_nrealargs 0) (mind_nrealdecls 0)
>   (mind_kelim (InProp InSet InType))
>   (mind_nf_lc
>     ((Prod (Name (Id smynat))
>            (Ind (((Mutind (MPfile (DirPath ((Id Datatypes) (Id Init) (Id Coq))))
>                 (DirPath ()) (Id nat)) 0) (Instance ())))
>            (Rel 2))))
>   (mind_consnrealargs (1))
>   (mind_consnrealdecls (1))
>   (mind_recargs (RTree
>           (Mrec ((Mutind (MPfile (DirPath ((Id SerTop)))) (DirPath ()) (Id snatbox)) 0)) 1))
>   (mind_nb_constant 0)
>   (mind_nb_args 1)
>   (mind_reloc_tbl ((1 1))))))
>  (mind_record (()))
>  (mind_finite BiFinite)
>  (mind_ntypes 1)
>  (mind_hyps ())
>  (mind_nparams 0)
>  (mind_nparams_rec 0)
>  (mind_params_ctxt ())
>  (mind_polymorphic false)
>  (mind_universes ((Instance ()) ()))
>  (mind_private ())
>  (mind_typing_flags ((check_guarded true) (check_universes true))))

And the function in charge of printing is `Printmod.pr_mutual_inductive_body`:

> if mib.mind_record <> None && not !Flags.raw_print then
>    print_record env mind mib
>  else
>    print_mutual_inductive env mind mib

I guess the `raw_print` flag is to blame here.

However, after having a look to how things are printed, it seems to me that the representation of records/inductives in the kernel is a bit weird, isn't it? Are they sharing the constructors field for both projections and constructors?

## Bugzilla Comment ID: 15081
Date: 2017-01-10 22:58:35 +0100
From: @_ejgallego

So indeed you are right, the documentation of Set Printing All is incorrect.

The weird printing is due to the internal representation, if the _ in the product wouldn't be inserted I guess we'd see:

> Variant snatbox : Set :=  Snatbox : forall smynat : nat, snatbox

## Bugzilla Comment ID: 15082
Date: 2017-01-10 23:00:53 +0100
From: @_ejgallego

This also explains the Variant/Record difference. In `print_mutual_inductive`:

>    match mib.mind_finite with
>    | Finite -> "Inductive"
>    | BiFinite -> "Variant"
>    | CoFinite -> "CoInductive"

vs in `print_record`:

>    match mib.mind_finite with
>    | BiFinite -> "Record"
>    | Finite -> "Inductive"
>    | CoFinite -> "CoInductive"

## Bugzilla Comment ID: 15083
Date: 2017-01-10 23:11:59 +0100
From: @_ejgallego

Regarding some other comments, I arrived at b18b40878f071b6c7d67d1a2d031370f7a498d0b , but need to understand better what is going on.

## Bugzilla Comment ID: 15124
Date: 2017-01-16 15:52:52 +0100
From: @_herbelin

Hi,

Internally (in the kernel), a type can be BiFinite, Finite, CoFinite indicating whether the type is meant to be not recursive, not recursive with well-founded terms, or not recursive with well-founded destructors respectively. Internally, the kernel has also a flag is_record to tell if the type is defined from projections and if yes if the projections are primitive.

My understanding is that the idea of Arnaud was that BiFinite types should be defined in an exclusive way using either Record/Structure when from fields (i.e. with { ... }), or Variant when from constructors (i.e. with C:type), so that Record/Structure, Variant, Inductive and CoInductive covers a distinct space of the different possibilities.

Variant: BiFinite, not record (necessarily not using { ... })
Record/Structure: BiFinite, record (necessarily using { ... })
Inductive: Finite, record if with { ... }, otherwise no

Following this idea, it seems to me that what should be addressed is the following:

> Structure sfoo := sbar : sfoo.

should fail as well

> Fail Record rnatbox := rmynat : nat.
> Fail Record rfoo := rbar : rfoo.
> (* --> Error: The Record keyword cannot be used to define a variant type.
>   Use Variant instead. *)

I would suggest something like "Error: The Record keyword is for types defined  using the syntax { ... }", (or even a syntax error). 

Then, function vernac_inductive would eventually happily welcome a simplification (and the kernel maybe accept mutual inductive types made from projections, or even made from a mix of types defined from projections and types not made from projections). For instance, why not to eventually accept "Inductive A := {a : B} with B := I : A"?

About printing, I don't know what is good and I did not do anything about Set Printing All.

After all, since the introduction of syntax such as

Inductive L := { hd : nat ; tl : L }.
Coinductive L := { hd : nat ; tl : L }.

there would be no reason not to also accept

Variant L := { p : nat }.

so that the Record and Structure keywords are not any more primitively needed per se.

So, in short, I fixed the most evident problems (inconsistency Record/Structure and unclear error message), but not more.

In particular, the code of print_record and pr_mutual_inductive is correct wrt my understanding of Arnaud's intent (even if maybe a different model allowing { ... } in Variant and strictly seeing Record and Structure as redundant is also possible).
